File: roms/star_force.py
#
#	Star Force
#

from base64 import b64encode
from io import BytesIO
from math import ceil
from sys import argv
from zipfile import ZipFile
from PIL import Image

with ZipFile(argv[1]) as z:
    prg1 = z.read('3.3p') + z.read('2.3mn')
    prg2 = z.read('1.3hj')
    fg = z.read('7.2fh') + z.read('8.3fh') + z.read('9.3fh')
    bg1 = z.read('15.10jk') + z.read('14.9jk') + z.read('13.8jk')
    bg2 = z.read('12.10de') + z.read('11.9de') + z.read('10.8de')
    bg3 = z.read('18.10pq') + z.read('17.9pq') + z.read('16.8pq')
    obj = z.read('6.10lm') + z.read('5.9lm') + z.read('4.8lm')
    snd = z.read('07b.bin')

rom = prg1 + prg2 + fg + bg1 + bg2 + bg3 + obj + snd

def pngstring(a):
    w = 1024
    img = Image.new('P', (w, ceil(len(a) / w)))
    img.putpalette(sum([[i, 0, 0] for i in range(256)], []))
    img.putdata(a)
    buf = BytesIO()
    img.save(buf, 'PNG')
    return b64encode(buf.getvalue())

with open(argv[2], 'wb') as f:
    f.write(b'export const ROM = \'data:image/png;base64,' + pngstring(rom) + b'\';\n')

File: roms/star_force.js
/*
 *	Star Force
 */

const fs = require('fs');
const {createCanvas, createImageData} = require('canvas');
const {Zlib} = require('zlibjs/bin/unzip.min');
Zlib.Unzip.prototype.read = Zlib.Unzip.prototype.decompress;

const cat = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0));
	return args.reduce((a, b) => (array.set(b, a), a + b.length), 0), array;
};

const z = new Zlib.Unzip(fs.readFileSync(process.argv[2]));
const prg1 = cat(z.read('3.3p'), z.read('2.3mn'));
const prg2 = z.read('1.3hj');
const fg = cat(z.read('7.2fh'), z.read('8.3fh'), z.read('9.3fh'));
const bg1 = cat(z.read('15.10jk'), z.read('14.9jk'), z.read('13.8jk'));
const bg2 = cat(z.read('12.10de'), z.read('11.9de'), z.read('10.8de'));
const bg3 = cat(z.read('18.10pq'), z.read('17.9pq'), z.read('16.8pq'));
const obj = cat(z.read('6.10lm'), z.read('5.9lm'), z.read('4.8lm'));
const snd = z.read('07b.bin');

const rom = cat(prg1, prg2, fg, bg1, bg2, bg3, obj, snd);

const pngString = array => {
	const w = 1024, h = Math.ceil(array.byteLength / w), canvas = createCanvas(w, h);
	const palette = new Uint8ClampedArray(new Uint32Array(256).map((e, i) => i | 0xff000000).buffer);
	canvas.getContext('2d', {pixelFormat: 'A8'}).putImageData(createImageData(new Uint8ClampedArray(array.buffer), w, h), 0, 0);
	return Buffer.from(canvas.toBuffer('image/png', {compressionLevel: 9, filters: canvas.PNG_FILTER_NONE, palette})).toString('base64');
};

fs.writeFileSync(process.argv[3], `export const ROM = 'data:image/png;base64,${pngString(rom)}';\n`);
